### PR TITLE
  Add new config option for DynamicHostPortRange 

### PR DESCRIPTION
**Pull request by SreeeS**
### Summary

Added a new config option to configure ephemeral host port range

### Implementation details

1. Added a new config option to configure ephemeral host port range by adding a new environment variable to the config struct which would be called from the `NewConfig` function. 
2. Modified the current fallback logic in the case of no host port range and added this fallback logic to work while parsing the environment variable for `DynamicHostPortRange`.  
3. Added the unit tests for the changes

### Testing

1. Tested the new dynamic host port range config option by running the `taskDefinitions` and checking the `portbindings` for the `hostPortRange` by describing the tasks.

- Case1: Passed the wrong dynamic host port range value and received the default value.
- Case2:  Passed the expected dynamic host port range value and received the `hostPortRange` with in user defined `DynamicHostPortRange`
- Case3: Passed no value and still received the default value.

2. Pass the unit tests in the GitHub workflow

New tests cover the changes: yes

### Description for the changelog

Enhancement: added new agent configuration to specify ephemeral host port range for port range mapping.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/config/types.go [hunk 1]
```diff
@@ -29,10 +29,12 @@ type ImagePullBehaviorType int8
 type ContainerInstancePropagateTagsFromType int8
 
 type Config struct {
+
```
Comment (singholt) [inline agent/config/types.go:32]: nit: remove this new line

File: agent/config/types.go [hunk 2]
```diff
@@ -359,4 +361,9 @@ type Config struct {
 	// WarmPoolsSupport specifies whether the agent should poll IMDS to check the target lifecycle state for a starting
 	// instance
 	WarmPoolsSupport BooleanDefaultFalse
+
+	// PortRange Variable enables the clients to set their custom port range
```
Comment (singholt) [inline agent/config/types.go:365]: please update this comment block to reflect the var name, something like

```
// DynamicHostPortRange specifies the dynamic host port range that the agent 
// uses to assign host ports for a container port range mapping.
// This defaults to the platform specific ephemeral host port range
```

File: agent/config/config_test.go [hunk 3]
```diff
@@ -123,6 +123,7 @@ func TestEnvironmentConfig(t *testing.T) {
 	)
 
 	defer setTestRegion()()
+	defer setTestEnv("ECS_DYNAMIC_HOST_PORT_RANGE", "200-300")()
```
[the remark this answers is not in the document]
Comment (SreeeS) [reply agent/config/config_test.go:126]: Sure, done.

File: agent/config/types.go (hunk 1 above)
Comment (SreeeS) [reply agent/config/types.go:32]: Done

File: agent/config/types.go (hunk 2 above)
Comment (SreeeS) [reply agent/config/types.go:365]: Sure, updated it.

File: agent/config/parse.go [hunk 4]
```diff
@@ -23,8 +23,10 @@ import (
 	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/dockerclient"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
```
[the remark this answers is not in the document]
Comment (SreeeS) [reply agent/config/parse.go:26]: Sure, done.

File: agent/config/parse.go [hunk 5]
```diff
@@ -372,3 +374,22 @@ func parseCgroupCPUPeriod() time.Duration {
 
 	return defaultCgroupCPUPeriod
 }
+
+var getDynamicHostPortRange = utils.GetDynamicHostPortRange
+
+func parseDynamicHostPortRange(dynamicHostPortRangeEnv string) string {
```
[the remark this answers is not in the document]
Comment (SreeeS) [reply agent/config/parse.go:380]: Sure, I have added it!

File: README.md [hunk 6]
```diff
@@ -120,91 +120,92 @@ The table below provides an overview of optional environment variables that can
 [the Amazon ECS developer guide](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-agent-config.html) for
 additional details on each available environment variable.
 
-| Environment Key | Example Value(s)            | Description | Default value on Linux | Default value on Windows |
```
[the remark this answers is not in the document]
Comment (SreeeS) [reply README.md]: Made changes to README file to not change the whole table but only the line I have added.

File: agent/config/config_test.go [hunk 7]
```diff
@@ -494,6 +498,62 @@ func TestValidFormatParseEnvVariableDuration(t *testing.T) {
 	assert.Equal(t, 1*time.Second, duration, "Unexpected value parsed in parseEnvVariableDuration.")
 }
 
+func TestValidFormatParseDynamicHostPortRange(t *testing.T) {
```
Comment (singholt) [inline agent/config/config_test.go:501]: nit: I see that the test has multiple cases now (both success and failure), so `TestParseDynamicHostPortRange` seems more suited.

**Review by singholt — APPROVED** [review]
🚀

File: agent/utils/ephemeral_ports_linux.go [hunk 8]
```diff
@@ -23,17 +23,17 @@ import (
...
 
 const (
-	// defaultPortRangeStart indicates the first port in ephemeral port range
-	defaultPortRangeStart = 49153
-	// defaultPortRangeEnd indicates the last port in ephemeral port range
-	defaultPortRangeEnd = 65535
 	// portRangeKernelParam is a kernel parameter that defines the ephemeral port range
 	portRangeKernelParam = "/proc/sys/net/ipv4/ip_local_port_range"
+	// DefaultPortRangeStart indicates the first port in ephemeral port range
+	DefaultPortRangeStart = 49153
```
Comment (prateekchaudhry) [inline agent/utils/ephemeral_ports_linux.go:29]: nit: Understanding that in Linux systems, ephemeral ranges are decided by ```portRangeKernelParam```. Does this value have 49153 have a significance for such cases?

**Review by prateekchaudhry — DISMISSED** [review]

File: agent/utils/ephemeral_ports_linux.go (hunk 8 above)
Comment (SreeeS) [reply agent/utils/ephemeral_ports_linux.go:29]: There is a possibility of getting an error while trying to use `portRangeKernelParam` when we call `GetDynamicHostPortRange` from parse.go which will then fallback to these default values.

**Review by singholt — APPROVED** [review]

**Review by prateekchaudhry — APPROVED** [review]

**Pull request merged**